### PR TITLE
Darwin: Allow certificates table to read DER/PEM files

### PR DESCRIPTION
**Pull request by russellhancox**

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/tables/system/darwin/certificates.mm [hunk 1]
```diff
@@ -66,17 +51,92 @@ void genCertificate(const SecCertificateRef& SecCert, QueryData& results) {
...
+
+    // If cert couldn't be read as DER, attempt
+    // to read it as a PEM.
+    if (cert == nullptr) {
+      BIO* bio = BIO_new_mem_buf((void*)bytes, content.size());
+
+      // There might be multiple certificates in the PEM.
+      while (1) {
+        PEM_read_bio_X509(bio, &cert, nullptr, nullptr);
+        if (cert == nullptr) break;
```
Comment (sharvilshah) [inline osquery/tables/system/darwin/certificates.mm]: Prefer braces with conditionals, even if they are a single line.

Comment (russellhancox) [inline osquery/tables/system/darwin/certificates.mm]: Done. clang-format moves the break to a new line but doesn't add braces.

**Comment by osqueryer** [comment]
Build origin/pr/1914/merge (Job: 2457) is good! :+1:

File: osquery/tables/system/darwin/certificates.mm [hunk 2]
```diff
@@ -18,24 +18,9 @@
 namespace osquery {
 namespace tables {
 
-void genCertificate(const SecCertificateRef& SecCert, QueryData& results) {
+void genCertificate(X509* cert, const std::string path, QueryData& results) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/certificates.mm]: We should pass `path` by reference:
```
void genCertificate(X509* cert, const std::string& path, QueryData& results)
```

The alternative is to copy (non-const) then `std::move` (invoke move assignment) on line 37. But IMO they are the same, and a pass-by-const-reference is the cleaner and easier to understand approach.

File: osquery/tables/system/darwin/certificates.mm [hunk 3]
```diff
@@ -66,17 +51,94 @@ void genCertificate(const SecCertificateRef& SecCert, QueryData& results) {
...
+    X509_free(cert);
+  }
+
   CFRelease(der_encoded_data);
 }
 
+void genFileCertificate(const std::string& path, QueryData& results) {
+  std::string content;
+  auto s = readFile(path, content);
+  if (s.ok()) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/certificates.mm]: Alternatively:
```
if (!s.ok()) {
  return;
}
```

File: osquery/tables/system/darwin/certificates.mm [hunk 4]
```diff
@@ -66,17 +51,94 @@ void genCertificate(const SecCertificateRef& SecCert, QueryData& results) {
...
+    const unsigned char* bytes = (const unsigned char*)content.c_str();
+    X509* cert = d2i_X509(nullptr, &bytes, content.size());
+
+    // If cert couldn't be read as DER, attempt
+    // to read it as a PEM.
+    if (cert == nullptr) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/certificates.mm]: What happens if `cert` **could** be read as DER? Do you want to do anything with it? In any case, it should be `X509_free`d, right?

File: osquery/tables/system/darwin/certificates.mm [hunk 5]
```diff
@@ -66,17 +51,94 @@ void genCertificate(const SecCertificateRef& SecCert, QueryData& results) {
...
+      BIO* bio = BIO_new_mem_buf((void*)bytes, content.size());
+
+      // There might be multiple certificates in the PEM.
+      while (1) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/certificates.mm]: I don't understand this. Is the `BIO` structure updated after the call to `PEM_read_bio_X509`?

I'm also very scared of `while(1)`, is it possible to refactor this a bit to include the call to `PEM_read_bio_x509` in the conditional? Or have a max number of attempts, so in the super weird cases this does not lead to a hang.

File: osquery/tables/system/darwin/certificates.mm (hunk 4 above)
Comment (russellhancox) [inline osquery/tables/system/darwin/certificates.mm]: I'd changed the logic to handle multiple certs in a PEM and accidentally deleted the lines that handle DER certs.

File: osquery/tables/system/darwin/certificates.mm (hunk 5 above)
Comment (russellhancox) [inline osquery/tables/system/darwin/certificates.mm]: Yes, the `BIO` structure keeps track of where it is.

I hadn't realized `PEM_read_bio_X509` also returns a pointer to a cert, so I've changed the for loop to be based on that. Let me know if you'd still like a limit.

**Comment by osqueryer** [comment]
Build origin/pr/1914/merge (Job: 2459) is good! :+1:

File: osquery/tables/system/darwin/keychain_utils.cpp [hunk 6]
```diff
@@ -121,9 +121,14 @@ void genAlgorithmProperties(X509* cert,
   }
 }
 
-std::string genSHA1ForCertificate(const CFDataRef& raw_cert) {
-  return hashFromBuffer(
-      HASH_TYPE_SHA1, CFDataGetBytePtr(raw_cert), CFDataGetLength(raw_cert));
+std::string genSHA1ForCertificate(X509* cert) {
+  const EVP_MD* fprint_type = EVP_sha1();
+  unsigned char fprint[EVP_MAX_MD_SIZE];
```
Comment (theopolis) [inline osquery/tables/system/darwin/keychain_utils.cpp]: Sorry I didn't catch this before:
```cpp
unsigned char fprint[EVP_MAX_MD_SIZE] = {0};
```

File: osquery/tables/system/darwin/keychain_utils.cpp [hunk 7]
```diff
@@ -121,9 +121,14 @@ void genAlgorithmProperties(X509* cert,
...
+  unsigned int fprint_size;
```
Comment (theopolis) [inline osquery/tables/system/darwin/keychain_utils.cpp]: Initialize to `0` I'm guessing?

File: osquery/tables/system/darwin/keychain_utils.cpp [hunk 8]
```diff
@@ -121,9 +121,14 @@ void genAlgorithmProperties(X509* cert,
...
+
+  X509_digest(cert, fprint_type, fprint, &fprint_size);
```
Comment (theopolis) [inline osquery/tables/system/darwin/keychain_utils.cpp]: Perhaps check the result:
```cpp
if (X509_digest(cert, fprint_type, fprint, &fprint_size)) {
  return genKIDProperty(fprint, fprint_size);
}
return "";
```

> X509_digest converts the X509 structure a to DER-encoded form; computes the message digest of it, using as the message digest algorithm; and returns the results in md, with the length of the returned digest in len. 1 is returned on success, or 0 on error.

Not sure if it'd catch/prevent any errors, but worth being clear to only operate on `fprint` if we expect to have it filled int with meaningful data.

File: osquery/tables/system/darwin/certificates.mm [hunk 9]
```diff
@@ -66,17 +51,94 @@ void genCertificate(const SecCertificateRef& SecCert, QueryData& results) {
...
+  if (cert != nullptr) {
+    genCertificate(cert, path, results);
+    X509_free(cert);
+  } else {
+    // If cert couldn't be read as DER, attempt
+    // to read it as a PEM.
+    BIO* bio = BIO_new_mem_buf((void*)bytes, content.size());
+
+    // There might be multiple certificates in the PEM.
+    while (PEM_read_bio_X509(bio, &cert, nullptr, nullptr) != nullptr) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/certificates.mm]: Yeah! This reads much better now! :dancer:

File: osquery/tables/system/darwin/certificates.mm [hunk 10]
```diff
@@ -66,17 +51,94 @@ void genCertificate(const SecCertificateRef& SecCert, QueryData& results) {
...
+        }
+        return status;
+      }));
+
+  if (!paths.empty()) {
+    for (const auto& path : paths) {
+      SecKeychainRef keychain = nullptr;
+      SecKeychainStatus keychain_status;
+      auto status = SecKeychainOpen(path.c_str(), &keychain);
+      if (status == 0 && keychain != nullptr &&
```
Comment (theopolis) [inline osquery/tables/system/darwin/certificates.mm]: I'm worried about this combo of conditionals. Do you mind breaking them up into subsequent checks?

Something like:
```cpp
if (status != errSecSuccess || keychain == nullptr || SecKeychainGetStatus(keychain, &keychain_status) != errSecSuccess) {
  if (keychain != nullptr) {
    CFRelease(keychain);
  }
  continue;
}
keychain_paths.insert(path);
```

works too.

We've seen crazy stuff where the return status indicates error but the output parameter still contains a result pointer for which the caller is responsible for-- combining a check for status + `nullptr` means a memory leak. Also, if `SecKeychainOpen` returns success, then `SecKeychainGetStatus` fails, there will be a memory leak.

Comment (russellhancox) [inline osquery/tables/system/darwin/certificates.mm]: This does read much better

**Comment by osqueryer** [comment]
Build origin/pr/1914/merge (Job: 2463) failed :-1:

**Comment by russellhancox** [comment]
It did, should be better now.

**Comment by theopolis** [comment]
Perfecto!

**Comment by osqueryer** [comment]
Build origin/pr/1914/merge (Job: 2464) failed :-1:

**Pull request merged**